### ISSUE TITLE
bug report: SELinux incompatibility using named volumes

### ISSUE THREAD

**Issue by robbertkl** [docker-mailserver/docker-mailserver#3887]
### 📝 Preliminary Checks

- [X] I tried searching for an existing issue and followed the [debugging docs](https://docker-mailserver.github.io/docker-mailserver/latest/config/debugging/) advice, but still need assistance.

### 👀 What Happened?

On my FedoraCoreOS host, SELinux is enabled ("enforcing") by default, and I run into some issues with docker-mailserver on my setup, using named volumes (see config below). Initial startup is good, but when the container is recreated (e.g. `docker compose up -d --force-recreate mail`) it breaks (see logs below).

I understand this might be out of scope as it concerns SELinux, but would still like to check if this is not somehow an issue with the image, as it only affects 1 of the volumes (`mail-state`).

Workaround where it doesn't break:
- Disabling SELinux system wide (`sudo setenforce 0`)
- Running the container privileged (`privileged: true` in compose.yaml)
- Disabling SELinux for the container (`security_opt: - label:disable` in compose.yaml)

Interestingly, adding bind option `:z` or `:Z` to the mounts (which is the suggested solution in many places) does not help; it looks like it's not used/allowed for named volumes (only bind mounts) which is why compose warns about that in some of its commands.

Other than disabling SELinux I don't have a solution for this. I can see the directories within `mail-state` are improperly labeled and I also cannot access them using docker exec. However, other volumes like `mail-data` also get some automatically created dirs, but they are labeled `system_u:object_r:container_file_t:s0` and can be accessed without issues:

```sh
# ls -Z /var/mail /var/mail-state | cat
/var/mail:
system_u:object_r:container_file_t:s0 example.com

/var/mail-state:
system_u:object_r:container_file_t:s0:c1022,c1023 lib-dovecot
system_u:object_r:container_file_t:s0:c1022,c1023 lib-fail2ban
system_u:object_r:container_file_t:s0:c1022,c1023 lib-logrotate
system_u:object_r:container_file_t:s0:c1022,c1023 lib-postfix
system_u:object_r:container_file_t:s0:c1022,c1023 lib-redis
system_u:object_r:container_file_t:s0:c1022,c1023 lib-rspamd
system_u:object_r:container_file_t:s0:c1022,c1023 spool-postfix
```

Anything special about the way mail-state is initially provisioned that makes it behave in this way? Any clue why I don't have this issue with the other named volumes?

### 🐋 DMS Version

v13.3.1

### 💻 Operating System and Architecture

FedoraCoreOS 39 x86_64

### ⚙️ Container configuration files

```yml
volumes:
  # ...
  letsencrypt:
    name: letsencrypt
  mail-data:
    name: mail-data
  mail-state:
    name: mail-state
  mail-logs:
    name: mail-logs
  mail-config:
    name: mail-config

services:
  # ...
  mail:
    container_name: mail
    image: mailserver/docker-mailserver
    hostname: ######
    restart: always
    cap_add:
      - NET_ADMIN
    ports:
      - '25:25'
      - '465:465'
      - '587:587'
      - '993:993'
    volumes:
      - mail-data:/var/mail
      - mail-state:/var/mail-state
      - mail-logs:/var/log/mail
      - mail-config:/tmp/docker-mailserver
      - letsencrypt:/etc/letsencrypt
    environment:
      ENABLE_RSPAMD: 1
      ENABLE_FAIL2BAN: 1
      ENABLE_CLAMAV: 0
      ENABLE_OPENDMARC: 0
      ENABLE_OPENDKIM: 0
      ENABLE_POLICYD_SPF: 0
      ENABLE_AMAVIS: 0
      ENABLE_SPAMASSASSIN: 0
      ENABLE_UPDATE_CHECK: 0
      DOVECOT_MAILBOX_FORMAT: mdbox
      SSL_TYPE: letsencrypt
      SPOOF_PROTECTION: 1
      MOVE_SPAM_TO_JUNK: 1
      MARK_SPAM_AS_READ: 1
      RSPAMD_LEARN: 1
      TZ: Europe/Amsterdam
```

### 📜 Relevant log output

```Text
[   INF   ]  Welcome to docker-mailserver v13.3.1
[   INF   ]  Checking configuration
[   INF   ]  Configuring mail server
chown: cannot read directory '/var/mail-state/lib-rspamd': Permission denied
chown: cannot read directory '/var/mail-state/lib-redis': Permission denied
chown: cannot read directory '/var/mail-state/lib-logrotate': Permission denied
chown: cannot read directory '/var/mail-state/lib-postfix': Permission denied
chown: cannot read directory '/var/mail-state/spool-postfix': Permission denied
chown: changing ownership of '/var/mail-state/spool-postfix': Permission denied
chgrp: cannot access '/var/mail-state/spool-postfix/maildrop': Permission denied
chgrp: cannot access '/var/mail-state/spool-postfix/public': Permission denied
chmod: cannot access '/var/mail-state/spool-postfix/maildrop': Permission denied
chmod: cannot access '/var/mail-state/spool-postfix/public': Permission denied
rm: cannot remove '/var/mail-state/spool-postfix/dev': Permission denied
rm: cannot remove '/var/mail-state/spool-postfix/etc': Permission denied
rm: cannot remove '/var/mail-state/spool-postfix/lib': Permission denied
rm: cannot remove '/var/mail-state/spool-postfix/pid': Permission denied
rm: cannot remove '/var/mail-state/spool-postfix/usr': Permission denied
rm: cannot remove '/var/mail-state/spool-postfix/private/auth': Permission denied
[   INF   ]  Starting daemons
[   INF   ]  ###### is up and running
Feb 12 21:36:12 mail postfix[858]: fatal: chdir(/var/spool/postfix): Permission denied
Feb 12 21:36:17 mail postfix[983]: fatal: chdir(/var/spool/postfix): Permission denied
Feb 12 21:36:22 mail postfix[1086]: fatal: chdir(/var/spool/postfix): Permission denied
...
```

**Comment by polarathene** [comment]
> I can see the directories within `mail-state` are improperly labeled and I also cannot access them using docker exec. However, other volumes like `mail-data` also get some automatically created dirs, but they are labeled `system_u:object_r:container_file_t:s0` and can be accessed without issues

`mail-state` is managed entirely by [this script](https://github.com/docker-mailserver/docker-mailserver/blob/22555347edeccd237fd1327a9cb09426b71e9679/target/scripts/startup/setup.d/mail_state.sh) during startup.

Does the issue happen on the initial container instance + volume creation? Or only after a restart?

Named volumes do differ from bind mounts in a few ways. I recall that they'll inherit any data at the internal mount location of the image, while bind mounts replace the contents (_additionally removing somewhat important attributes from the mount path like with `/var/mail` with SGID bit, along with ownership_).

No clue how SELinux further complicates that. But perhaps your experience with it and looking at the related script file can identify the issue. Basically we copy the data from it's original location in `/var/lib/` and change the original location to symlink to the new central location where only runtime state relevant to DMS is kept.

I don't know enough about SELinux to provide any additional insights for why it's forbidding you from those operations (_at least at that location_).

**Comment by robbertkl** [comment]
> `mail-state` is managed entirely by [this script](https://github.com/docker-mailserver/docker-mailserver/blob/22555347edeccd237fd1327a9cb09426b71e9679/target/scripts/startup/setup.d/mail_state.sh) during startup.

Great, I'll have a look!

> Does the issue happen on the initial container instance + volume creation? Or only after a restart?

No, not on the initial container, all is good there, but that's because the labels still match the running container. As soon as the volume is mounted on a new container (after a force-recreate "restart"), the directories are no longer accessible because they have the labels belonging to the original container.

> No clue how SELinux further complicates that. But perhaps your experience with it and looking at the related script file can identify the issue. Basically we copy the data from it's original location in `/var/lib/` and change the original location to symlink to the new central location where only runtime state relevant to DMS is kept.
> 
> I don't know enough about SELinux to provide any additional insights for why it's forbidding you from those operations (_at least at that location_).

Not really too familiar with SELinux myself, I'm just moving to a new host (OS) with SELinux enabled, which is why I'm running into these issues. Coming from a system without SELinux I'm OK with disabling it (or at least disabling it for this container), but even better to make it work in a SELinux-enforced setup out of the box.

It seems indeed the copying/moving from the running container makes the labels/context stick, in contrast with newly created (e.g. mkdir) directories like the initially created mail data structure in `/var/mail`. I see `cp` and `mv` have some SELinux related flags, so might be an easy fix for the script you mentioned. I'll have a look and do some testing and will submit a PR if it works.

**Comment by polarathene** [comment]
> Not really too familiar with SELinux myself, I'm just moving to a new host (OS) with SELinux enabled, which is why I'm running into these issues

It's been a concern for me with switching over to Fedora myself 😓

I guess I could just disable it, but it's not the only difference that adds friction (_my preferred distro is EndeavorOS / ArchLinux, mostly because it has minimal customizations/surprises to discover_).

I've encountered issues with SELinux a few times when I spun up a Fedora VPS/VM and tried using Docker and sometimes it was difficult to troubleshoot/understand the SELinux stuff. I remember having another issue just using rsync also failing due to SELinux.

> I'll have a look and do some testing and will submit a PR if it works.

That'd be greatly appreciated! ❤️

**Comment by georglauterbach** [comment]
Just my 2 cents here: [shouldn't such volumes be labeled with a `:Z` in Docker Compose](https://docker-mailserver.github.io/docker-mailserver/latest/usage/#configuration-steps) (/according to our docs)?

**Comment by robbertkl** [comment]
> Just my 2 cents here: [shouldn't such volumes be labeled with a `:Z` in Docker Compose](https://docker-mailserver.github.io/docker-mailserver/latest/usage/#configuration-steps) (/according to our docs)?

That's for bind mounts. It's not used/allowed for named volumes.

Actually it's not too clearly documented within the Docker / Compose docs, so it was the first thing I tried, but made no difference (tried both `:z` and `:Z`). For some compose commands (not all, strangely enough) it printed a warning that you can't specify bind options for volume mounts.

**Comment by georglauterbach** [comment]
I see, thanks :)
